Implementation of __cause__ for ValidationError using ExceptionGroups

zakstucke opened the pull request:
## Change Summary
<details>

<summary>Out of date, contents has changed significantly from initial draft. See the comments for the path to the current version.</summary>

**Draft wanting feedback**
- Fully working Python 3.7 upwards
- Requires [exceptiongroup](https://github.com/agronholm/exceptiongroup) dependency for `ExceptionGroup` backport pre 3.11
- Attaches causes of user made validation errors to the `__cause__` of a `ValidationError`, including the respective `loc` for each cause chain
- See below for new example traceback output

Following on from my [issue](https://github.com/pydantic/pydantic/issues/6498#issuecomment-1636790534) in pydantic main, if `ValidationError`(s) utilised `__cause__` they would become immensely more powerful in non-user facing, debugging scenarios. 

I've got a working but dirty implementation here, was trying for minimal lines. 

Before I spend more time on it making it pretty, safe and optimised, I'm looking for thoughts on whether this is something Pydantic is willing to support. 

This example implementation works with Python 3.7 upwards for compatibility, using the exceptiongroup backport to that effect. Works great in all python versions. (obviously `add_note()` is only used 3.11 upwards, but the chaining and groups are available on all supported versions).

For 3.11 upwards this could be improved to:
- Use `add_note()` on the main exception instead of the extra `UserWarningException` to specify the `loc`
- Not require the backport exception group lib

If you're not liking the bloat for the average new user, maybe this could be enabled with a global flag. 

This also seems to break 0 tests as nothing as of yet relies on `__cause__`.

**Edit: i've noticed something off with the second error group, aka the singular exception for `bar`, not sure why its not but should show the error location same as the source error of `foo` chain, would fix.**

</details>

<!-- Please give a short summary of the changes. -->
Implements an `ExceptionGroup` as the cause of a `ValidationError`, to display user-defined callback errors that caused the `ValidationError` aesthetically. Really useful for debugging as a user.

Pre 3.11 it uses the  [exceptiongroup](https://github.com/agronholm/exceptiongroup) backport, 3.11+ uses the rust-native `BaseExceptionGroup`. In both cases, the cause is created lazily on read, so no performance cost. For everything other than PyPy, the lazy cause is only built once.

The below example output is up to date.

## Related issue number
https://github.com/pydantic/pydantic/issues/6498#issuecomment-1636790534

<!-- Are there any issues opened that will be resolved by merging this change? -->
<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [x] Unit tests for the changes exist
* [ ] Documentation reflects the changes where applicable
* [x] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

```Python
from pydantic import BaseModel, field_validator

def check(v: int):
    if v < 0:
        raise ValueError("INNER ERR")
    return v

class Foo(BaseModel):
    foo: int
    bar: int

    @field_validator("foo")
    def check_foo(cls, v):
        if v < 0:
            try:
                check(v)
            except ValueError as e:
                new_err = ValueError("OUTER ERR")
                new_err.add_note("SOME EXTRA INFORMATION") # NOTE: 3.11+ only
                raise new_err from e
        return v
    
    @field_validator("bar")
    def check_bar(cls, v):
        if v < 0:
            check(v)
        return v    
    
Foo(foo=-1, bar=-1)
```

3.11+
```
  | ExceptionGroup: Pydantic User Code Exceptions (2 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 51, in check_foo
    |     check(v)
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 40, in check
    |     raise ValueError("INNER ERR")
    | ValueError: INNER ERR
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | Traceback (most recent call last):
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 55, in check_foo
    |     raise new_err from e
    | ValueError: OUTER ERR
    | SOME EXTRA INFORMATION
    | 
    | Pydantic: cause of loc: foo
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 61, in check_bar
    |     check(v)
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 40, in check
    |     raise ValueError("INNER ERR")
    | ValueError: INNER ERR
    | 
    | Pydantic: cause of loc: bar
    +------------------------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "xxxxxxxxxxx/pydantic-core/test.py", line 64, in <module>
    Foo(foo=-1, bar=-1)
  File "xxxxxxxxxxx/pydantic-core/venv/lib/python3.11/site-packages/pydantic/main.py", line 150, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 2 validation errors for Foo
foo
  Value error, OUTER ERR [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
bar
  Value error, INNER ERR [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
```

Pre 3.11 & PyPy:
```
  | exceptiongroup.ExceptionGroup: Pydantic User Code Exceptions (2 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 16, in check_foo
    |     check(v)
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 5, in check
    |     raise ValueError("INNER ERR")
    | ValueError: INNER ERR
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | Traceback (most recent call last):
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 20, in check_foo
    |     raise new_err from e
    | ValueError: OUTER ERR
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | UserWarning: Pydantic: cause of loc: foo
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 26, in check_bar
    |     check(v)
    |   File "xxxxxxxxxxx/pydantic-core/test.py", line 5, in check
    |     raise ValueError("INNER ERR")
    | ValueError: INNER ERR
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | UserWarning: Pydantic: cause of loc: bar
    +------------------------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "xxxxxxxxxxx/pydantic-core/test.py", line 29, in <module>
    Foo(foo=-1, bar=-1)
  File "xxxxxxxxxxx/pydantic-core/venv10/lib/python3.10/site-packages/pydantic/main.py", line 150, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 2 validation errors for Foo
foo
  Value error, OUTER ERR [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
bar
  Value error, INNER ERR [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
```

Selected Reviewer: @dmontagu

adriangb commented:
I think this is an interesting idea!

I haven't looked in detail at how it works, but I will say that what really should happen in the end is that `ValidationError` should be itself an `ExceptionGroup`. It's just not trivial to make that happen due to a combination of the state of the current codebase, limitations of multiple inheritance with PyO3/lack of an ExceptionGroup C API and performance of calling into Python (like you are doing in this PR).

I will look in more detail but just because of the calling into Python that's being done here I suspect there is going to be a huge performance hit to benchmarks (and it looks like CodSpeed is already flagging `test_raise_error_value_error` as problematic) that means we can't move forward with this as-is :(

zakstucke commented:
@adriangb if the performance of the backport is the problem, this could be supported 3.11+ only. 

I've updated the PR to only include the logic when python is 3.11+ (compile time excluded code on older versions), and uses the rust `PyBaseExceptionGroup`, so no python callbacks.

My focus is 3.11+, I personally wouldn't mind if this was a limitation.

However the arguments for using the backport in the older versions:
- This is a crazy fast function, this addition also doesn't run at all (aka zero cost) when the user isn't including python callback validators. This is why you see it affect so few benchmarks, so the cost of going from `0.0000562` to `0.000145` seconds (with the pre 3.11 backport version), given the user is already running callbacks to python, must be pretty negligble. 
- It can be immensely useful info, even if older versions have to manually enable with a global flag, I think a lot of people would want and enable it today if they could.

Out of interest, why would you eventually want the `ValidationError` itself to be an `ExceptionGroup`? To my knowledge you're only ever getting one validation error per rust call, which can have lots of children excs. So in my understanding having a singular validation exception, with an `ExceptionGroup` `__cause__` (that is built to house children excs, not a parent and children together) like this, is already as good as it gets with groups?

adriangb commented:
> Out of interest, why would you eventually want the ValidationError itself to be an ExceptionGroup? To my knowledge you're only ever getting one validation error per rust call (adding to the negligability of a small performance hit), which can have lots of children excs. So in my understanding having a singular validation exception, with an ExceptionGroup (that is built to house children excs, not a parent and children together) in the __cause__ like this, is already as good as it gets with groups?

A `ValidationError` is an error that represents a group of errors. That's what an ExceptionGroup is! There's also a tree-like relationship in errors that is currently present in the `loc` aspect of ValidationError but not in the error structure itself. That is, if you have a union and the union fails validation the cause of that failure is actually all of the failures that happened with the variants. So making `ValidationError` an ExceptionGroup would allow us to express that much better, in addition to enabling the 3.11+ syntax for filtering and traversing this tree when using `except`.

You can also do weird things (which I think this PR should consider btw) like having a python function validator which does:

```python
try:
    handler(x)
catch ValidationError as e:
    raise ValueError(...) from e
```

So now you have a ValidationError -> ValueError -> ValidationError relationship.

zakstucke commented:
Makes sense you're right. I guess it then comes down to the issues you'd mentioned with how infeasible it would be to convert `ValidationError` to an `ExceptionGroup` in the near future.

I've just tested out your edge case:
```Python
import typing as tp
from pydantic import BaseModel, field_validator, ValidationError, TypeAdapter, AfterValidator


def check(v: int):
    if v < 0:
        raise ValueError("INNER ERR")
    return v

sub_validator = TypeAdapter(tp.Annotated[int, AfterValidator(check)])

class Baz(BaseModel):
    bar: int

    @field_validator("bar")
    def check_bar(cls, v):
        try:
            return sub_validator.validate_python(v)
        except ValidationError as e:
            raise ValueError("SUB VALIDATION ERROR!") from e    

class Foo(BaseModel):
    foo: int
    baz: Baz    

    @field_validator("foo")
    def check_foo(cls, v):
        if v < 0:
            try:
                check(v)
            except ValueError as e:
                new_err = ValueError("OUTER ERR")
                new_err.add_note("SOME EXTRA INFORMATION")
                raise new_err from e
        return v
    
Foo(foo=-1, baz={"bar": -1})
```

Outputs:
```
  | ExceptionGroup: Pydantic User Code Exceptions (2 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "pydantic-core/test.py", line 31, in check_foo
    |     check(v)
    |   File "pydantic-core/test.py", line 7, in check
    |     raise ValueError("INNER ERR")
    | ValueError: INNER ERR
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | Traceback (most recent call last):
    |   File "pydantic-core/test.py", line 35, in check_foo
    |     raise new_err from e
    | ValueError: OUTER ERR
    | SOME EXTRA INFORMATION
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | UserWarning: Cause of ValidationError loc:  'foo'
    +---------------- 2 ----------------
    | ExceptionGroup: Pydantic User Code Exceptions (1 sub-exception)
    +-+---------------- 1 ----------------
      | ValueError: INNER ERR
      | 
      | The above exception was the direct cause of the following exception:
      | 
      | UserWarning: Cause of ValidationError loc:  'body'
      +------------------------------------
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | Traceback (most recent call last):
    |   File "pydantic-core/test.py", line 19, in check_bar
    |     return sub_validator.validate_python(v)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "pydantic-core/venv/lib/python3.11/site-packages/pydantic/type_adapter.py", line 206, in validate_python
    |     return self.validator.validate_python(__object, strict=strict, from_attributes=from_attributes, context=context)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | pydantic_core._pydantic_core.ValidationError: 1 validation error for function-after[check(), int]
  Value error, INNER ERR [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | Traceback (most recent call last):
    |   File "pydantic-core/test.py", line 21, in check_bar
    |     raise ValueError("SUB VALIDATION ERROR!") from e    
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | ValueError: SUB VALIDATION ERROR!
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | UserWarning: Cause of ValidationError loc:  'baz.bar'
    +------------------------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "pydantic-core/test.py", line 40, in <module>
    Foo(foo=-1, baz={"bar": -1})
  File "pydantic-core/venv/lib/python3.11/site-packages/pydantic/main.py", line 150, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 2 validation errors for Foo
foo
  Value error, OUTER ERR [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
baz.bar
  Value error, SUB VALIDATION ERROR! [type=value_error, input_value=-1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.0.3/v/value_error
```

Which works well out the box for the weird double `ValidationError` case, just creates a tree of groups.

lig commented on src/errors/validation_exception.rs:
```diff
@@ -62,9 +62,56 @@ impl ValidationError {
+                    Ok(err) => {
+                        let py_err = PyErr::from_value(err.into_ref(py));
+                        if cause.is_some() {
+                            py_err.set_cause(py, cause);
```
Is it possible to extract the code above into a function (`build_cause`?) that we will invoke here?

It looks like we could attempt building `cause` and won't use it in some cases. IMO, this would bring more clarity to the code as well.

zakstucke replied:
Moved to a func, it costs a `borrow()` keeping it all together, but negligble

zakstucke commented:
I'm close to happy with this now.  Fixed the random exception missing its traceback in the result and switched to a note rather than an extra warning exception.

The changes made to `convert_err()` & the `py_err_string!` macro are a bugfix to stop throwing away the top level traceback from user errors. Before this, an error raised in a validator has no traceback at all. 
It seems like it might be a Pyo3 quirk, but `PyErr.into_py(py)` seems to throw away the traceback. 

@davidhewitt do you have any idea why this is and if there's a better way to solve it than I've done? `PyErr.state` is private and there's no interface to attach a traceback / convert to `PyObject` without discarding it that I can see.

davidhewitt commented:
> It seems like it might be a Pyo3 quirk, but `PyErr.into_py(py)` seems to throw away the traceback.

Are you willing to reduce this to a minimal repro which can be reported as an issue to PyO3, and we can discuss there if it's a bug? Before 3.12 the Python exception state is stored as split type / value / traceback and so I'd be willing to believe the `PyErr::into_py()` implementation needs adjusting.

zakstucke commented:
@davidhewitt done :), if there's a cleaner way of getting around it on the current pyo3 version for the use here, please let me know.

davidhewitt commented:
Thanks - I think your implementation here is correct (and likely what we would land upstream in PyO3).

davidhewitt commented:
@adriangb and I just had a catch up about this PR. Overall we like this a lot and we're happy to offer users a way to get at inner exceptions.

Some thoughts which we had:
- Forwards compatibility with changing `ValidationError` to itself be an `ExceptionGroup`. This seems ok, we could make `ValidationError.__cause__` just be a reference to itself (a bit weird but would avoid breaking users who access `__cause__` as implemented in this PR.)
- Performance. We wondered whether we should add `#[getter] fn __cause__` to lazily create the `ExceptionGroup` rather than always create `__cause__` immediately. (We could cache it to create it just once.) Not a deal breaker but potentially nice to have.
- Feature set. If we create `__cause__` lazily we could also support <3.11 using the `exceptiongroup` backport. `__cause__` could throw on access if the backport isn't installed on those versions. What do you think about this?
- Structure of `__cause__`. This we'd like to make sure we're happy with before we merge this PR. Is it API breaking if we changed the contents of these exceptions later?
  - If we deem it _is_ API breaking to change the inner exceptions later, we need to be sure we're happy with them in various combinations of schemas / validators / etc., because it will impact our ability to iterate if users become highly sensitive to exception structure.
  - If it's not API breaking to change the inner exceptions later, then we're obviously fine to merge the structure as-is.

zakstucke commented:
@davidhewitt great to hear!

> Forwards compatibility ... we could make `ValidationError.__cause__` just be a reference to itself

No qualms against this for programmatic access, would definitely work.

> Performance ... lazy creation

TLDR it doesn't work. 

I'd actually already looked into this when trying to ease @adriangb's performance concerns with the backport. 
It would work for a user catching the exception and programatically acessing the `__cause__`, but for the main benefit of this, traceback crashing the process and having full context, that flow happens directly in CPython, `__cause__` isn't used in that case. See `print_exception_cause_and_context()` [here in cpython source](https://github.com/python/cpython/blob/main/Python/pythonrun.c#L1328) which is what I think happens; the order of logic in there meaning attaching the lazy creation to even something like `repr` doesn't seem to work.

If you've got any ideas on this let me know, I gave up. 

> Structure of `__cause__` / API break

Valid concern, given how much looks like needs improving in `ValidationError` I don't think it would be a good idea to guarantee a specific structure in `__cause__`. It's current intended use should be for visual context in stack traces, not programatic access. If a user starts searching through this error group for errors they're asking for trouble!

An option for making the above enforceable/covered pd side, we could add another method to the `ValidationError` api, `user_errs() -> list[Exception]`, which guarantees to produce a flat list of all user validator-caused errors inside the `ValidationError`. With the current implementation this would be just `return cause or []`, but would alleviate the worry of modifying `__cause__` down the line as everyone can be told to use that. 

So if lazy doesn't work, unless you guys think of something new it doesn't, options for backport are:
- No support, sad
- Full support, accept the 3x slowdown in ValidationError to py (on my machine 1us to 3us), but remember above comments about how imo this is neglible as the user would already been running python callbacks.
- Put behind feature flag, e.g. `pydantic_core.enable_user_err_context = True`

zakstucke commented:
@davidhewitt I have actually found a way to get around the lazy problem but it's a bit weird:

```Python
import sys

original_excepthook = sys.excepthook

def traverse_causes_on_crash(start_exc: BaseException):
    excs: list[BaseException] = [start_exc]
    if isinstance(start_exc, BaseExceptionGroup):
        excs.extend(start_exc.exceptions)

    for exc in excs:
        cause = getattr(exc, '__cause__', None)
        if cause is not None:
            traverse_causes_on_crash(cause)

def replacement_excepthook(exc_type, exc_value, exc_traceback):
    # Traverse the causes:
    traverse_causes_on_crash(exc_value)

    # Call the original excepthook:
    original_excepthook(exc_type, exc_value, exc_traceback)

if original_excepthook is not replacement_excepthook:
    sys.excepthook = replacement_excepthook
```

This allows the `__cause__` hydration to happen just before C checks the cause.
Is this too weird or something to keep looking into? 

But this does provide 0 cost for the backport version, could either just be run on `pydantic_core` import or enabled with `pydantic_core.enable_validation_cause_tracebacks()`. 

`traceback.print_exc()` etc already seems to work out the box (with lazy __cause__) when other libs/apps are catching errs and logging them themselves, so adding this would cover everything I think.

zakstucke commented:
@adriangb @davidhewitt 

Other than PyPy, this is now working on all versions with no performance cost, the `test-pydantic-integration` failure I believe is unrelated to this PR.

I'm more or less finished with it, so  its ready for you to let me know what you want changing. 

If you see what might be causing the segfault in PyPy (in the `__cause__` getter) let me know!

zakstucke commented:
PyPy fixed, it uses `__cause__` itself in `ffi::PyException_GetCause` & `ffi::PyException_SetCause` so was causing infinite recursion, disabled ffi setting for pypy. 

Converting from draft!

Please review

davidhewitt commented:
Thanks @zakstucke for the thorough round of work based on my thoughts above! Sorry it's taken me some time to return to this thread.

Your solution for my uninformed suggestion of lazy `__cause__` is very neat; I think on balance reading it through I regret the complexity which lazy `__cause__` has introduced. In particular I worry that relying on `sys.excepthook` will lead to weird edge cases when other users also are fiddling with `sys.excepthook`.

Having discussed this further with @samuelcolvin and @adriangb we'd like to suggest to go back to non-lazy `__cause__`. Additionally, we think we there should be a config flag for this as you suggested (though maybe I suggest the name `validation_error_cause` instead of `enable_user_err_context`), with the following documented properties:
- By default it's off (so no performance impact for those who don't need it). In future we may change default to on.
- We can document with the config that we make no promises for now about the structure of the contents of `__cause__`, to resolve our concern about API stability. (i.e. this is intended to help users debug, not for programmatic access)
- Enabling this flag on Python 3.10 or earlier without the `exceptiongroup` backport installed is an error. Whether it fails at model build time, or when trying to populate the `__cause__`, is down to whatever feels better for an implementation.

Please let me know if you'd like some clarification or detail on how any of that might be implemented. I'm aware of all the work you have already put into this PR; thank you, I'm aware I'm asking for yet more (hopefully for what should be a final implementation). If you don't have budget to continue on this I understand, I can get it over the line eventually. If you want to keep working on this, great and thanks!

zakstucke commented:
@davidhewitt all your points should now be covered! Sorry for the delay on this, life got in the way.

- Something like this is useful to be configured once globally, I couldn't see an existing good way to do that. Most of the bloat in here is adding a new `flags` object to complement `config` passed into `SchemaValidator` & `SchemaSerializer`.
    - New `CoreFlags` `TypedDict` is modelled off `CoreConfig`, currently only implementing `validation_error_cause`. The way this is setup though is that theoretically you could set `CoreFlags` to accept applicable configs from `CoreConfig`; when a variable has been set in `flags`, its value would be used if the value wasn't being overriden in the active `config/target`. 
    - The key difference between the two is `config` morphs as the object being validated is traversed, but the `flags` stay the same throughout.
    - The two objects are combined into a `UserConfig` reffed struct, when config values are accessed with `UserConfig::get_conf(py_key: &PyString)`, the value will be picked from `config/target`, falling back to `flags`, falling back to `None`. 

- `validation_error_cause` now obviously defaults to `false` when omitted.
- The backport isn't a dependancy, a runtime import error occurs if a user has enabled the flag <3.11 without installing the backport.
- After this has been implemented, a PR will be needed in pydantic to implement an interface to flags, so that it can be automatically added internally for calls to `TypeAdapter/BaseModel` etc, imo this would be done by a user something like:
```Python
from pydantic import flags
flags.set_validation_error_cause(True)
```
- There seems to be a small performance boost in error handling from this PR, which is always nice :)

Please review

davidhewitt commented on pyproject.toml:
```diff
@@ -1,7 +1,8 @@
     'maturin>=1,<2',
-    'typing-extensions >=4.6.0,!=4.7.0'
+    'typing-extensions >=4.6.0,!=4.7.0',
+    'exceptiongroup >=1.1.0'
```
I'm curious why we needed the backport as a build dependency?

zakstucke replied:
You're right its not needed, I'd added it here before I found `tests/requirements.txt`, removed!

davidhewitt commented on tests/requirements.txt:
```diff
@@ -11,3 +11,4 @@ pytest-pretty==1.2.0
 pytest-timeout==2.1.0
 pytz==2023.3
 numpy==1.25.2; python_version >= "3.9" and python_version < "3.12"
+exceptiongroup==1.1
```
Add python upper bound here perhaps?

```suggestion
exceptiongroup==1.1; python_version < "3.11"
```

zakstucke replied:
Added, modified test to not need it in 3.11

davidhewitt commented on src/errors/validation_exception.rs:
```diff
@@ -94,6 +115,91 @@ impl ValidationError {
+                    }
+                }
+
+                // Pre 3.11 notes support, use a UserWarning exception instead:
+                #[cfg(not(Py_3_11))]
+                {
+                    use pyo3::exceptions::PyUserWarning;
+
+                    let wrapped = PyUserWarning::new_err((note,));
+                    wrapped.set_cause(py, Some(PyErr::from_value(err.as_ref(py))));
+                    user_py_errs.push(wrapped);
+                }
```
I'm curious for the choice to wrap in a `UserWarning`?

zakstucke replied:
Pre 3.11 we can't use a note to specify the error location, so we need to do it in an error, I figured considering its not an error, a "warning" made the most sense, happy to change if you've got something better!

davidhewitt replied:
Given that pre-3.11 users won't be reading `__notes__`, think it's maybe better to just not emit notes for them? That way we don't emit a different exception type on different versions. (`ValueError` vs `UserWarning`)

davidhewitt reviewed (changes requested):
No problem, I know the feeling all too well.

The `flags` proposal is interesting but I think slightly orthogonal to the original motivation proposed here, would you mind please splitting off into a separate PR so we can discuss that in a new thread? (I also wonder whether that really needs core changes, `pydantic` may be able to just merge flags into the config. A workaround for this today is define a subclass of `BaseModel` with the config you want and use that everywhere.)

Some other small questions & comments posted inline, though I'll be able to give the new implementation a more thorough review when the flags-related changes are split off. 😄

zakstucke commented:
@davidhewitt addressed all your reviews and split the PR into two. I've gone into your comments on the flags in that PR: #897 

Please review

davidhewitt reviewed (commented):
Overall this looks good to me, thanks! I'd be glad for a final view from @adriangb or @samuelcolvin before we merge.

zakstucke commented:
@davidhewitt the note (3.11+) / `UserWarning` (<3.11) is used to attach the source of the error and is all about improving the traceback debugging experience.

If you check all the way to the top of this PR i give the examples for both versions. If we don't include the `UserWarning`  <3.11, these users have a worse debugging experience as they can't find exactly where in the model being analysed the error occurred, unlike all other `ValidationError`'s. 
Given we've already agreed stable programmatic access isn't currently being promised, I don't think there's an issue personally.

3.11+
![image](https://github.com/pydantic/pydantic-core/assets/44890343/8f5e464d-6fc0-436b-b81e-6d0a193b7fe7)

Pre:
![image](https://github.com/pydantic/pydantic-core/assets/44890343/badbe1e5-a6bc-4a2b-9fb2-d8b910303bae)

davidhewitt commented:
Thanks, yes that makes sense from a UX perspective and fits with our statement that this won't be stable.

samuelcolvin commented on tests/test_errors.py:
```diff
@@ -517,6 +521,137 @@ def test_all_errors():
+
+    enabled_config: CoreConfig = {'validation_error_cause': True}
+
+    def check_grouped_exception(exc: BaseException) -> BaseException:
```
better to have two tests and add add a `pytest.mark.skipif` to both, so only one runs.

zakstucke replied:
I've separated out the contents tests as you suggested here :)

samuelcolvin commented on tests/test_errors.py:
```diff
@@ -517,6 +521,137 @@ def test_all_errors():
+    assert repr(cause) == repr(ValueError('Oh no!'))
+    assert cause.__traceback__ is not None
+
+    def multi_raise_py_error(v: Any) -> Any:
```
this seems like a separate test.

samuelcolvin commented on tests/test_errors.py:
```diff
@@ -517,6 +521,137 @@ def test_all_errors():
+    with pytest.raises(ValidationError) as exc_info:
+        s2.validate_python('anything')
+
+    def validate_s2_chain(val_err: ValidationError):
```
better not to use functions like this as it becomes hard to understand upon failure, especially as it only removes one duplication.

samuelcolvin commented on tests/test_errors.py:
```diff
@@ -517,6 +521,137 @@ def test_all_errors():
+        def singular_raise_py_error(v: Any) -> Any:
+            raise ValueError('Oh no!')
+
+        for _, config, expected in config:
```
please use parameterize rather than a loop - loops in tests are impossible to understand/debug.

zakstucke replied:
Done!

samuelcolvin reviewed (commented):
otherwise LGTM.

Assuming @davidhewitt is happy with the rust code.

zakstucke commented:
@samuelcolvin @davidhewitt sorry for the delay, refactored the tests as requested :)

davidhewitt reviewed (approved):
Thanks again, this LGTM and it's nice that it delivers some neat performance improvements on the error pathway too! Sorry that this has had such long cycle times on my end.

zakstucke commented:
No problem and thanks! :tada:
